File: src/main/java/com/qtech/bubbles/entity/ammo/AmmoType.java
package com.qtech.bubbles.entity.ammo;

import com.qtech.bubbles.common.AttributeMap;
import com.qtech.bubbles.entity.AmmoEntity;
import com.qtech.bubbles.registry.RegistryEntry;

import java.awt.*;

public abstract class AmmoType extends RegistryEntry {
    public abstract void render(Graphics g, AmmoEntity rotation);

    @SuppressWarnings("EmptyMethod")
    public void onCollision() {
        // Do nothing
    }

    public abstract AttributeMap getDefaultAttributes();

    public abstract Shape getShape(AmmoEntity ammoEntity);
}

File: src/main/kotlin/qtech/bubbles/common/ammo/AmmoType.kt
package qtech.bubbles.common.ammo

import qtech.bubbles.common.AttributeMap
import qtech.bubbles.common.RegistryEntry
import qtech.bubbles.entity.AmmoEntity
import qtech.hydro.GraphicsProcessor
import java.awt.Shape

abstract class AmmoType : RegistryEntry() {
    abstract fun render(g: GraphicsProcessor, ammoEntity: AmmoEntity)
    fun onCollision() {
        // Do nothing
    }

    abstract val defaultAttributes: AttributeMap
    abstract fun getShape(ammoEntity: AmmoEntity): Shape
}